### PR TITLE
added sql db diagnostic logging enabled

### PR DESCRIPTION
**Pull request by muzzamilinovaqo**

### CODE REVIEW

File: plugins/azure/sqldatabases/dbDiagnosticLoggingEnabled.js [hunk 1]
```diff
@@ -0,0 +1,80 @@
+var async = require('async');
+var helpers = require('../../../helpers/azure');
+
+module.exports = {
+    title: 'Enable Diagnostic Logging for SQL Databases',
+    category: 'SQL Databases',
+    domain: 'Databases',
+    description: 'Enable diagnostic logging for enhanced monitoring and troubleshooting.',
```
Comment (fatima99s) [inline plugins/azure/sqldatabases/dbDiagnosticLoggingEnabled.js:8]: Ensures diagnostic logging is enabled for SQL Databases.

Comment (fatima99s) [inline plugins/azure/sqldatabases/dbDiagnosticLoggingEnabled.js:8]:
```suggestion
    description: 'Ensures diagnostic logging is enabled for SQL Databases.',
```

File: plugins/azure/sqldatabases/dbDiagnosticLoggingEnabled.js [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+                    } else {
+                        databases.data.forEach(database=> {
+                            
+                            var diagnosticSettings = helpers.addSource(cache, source, ['diagnosticSettings', 'listByDatabase', location, database.id]);
+                        
+                            if (!diagnosticSettings || diagnosticSettings.err || !diagnosticSettings.data) {
+                                helpers.addResult(results, 3, 'Unable to query diagnostic settings: ' + helpers.addError(diagnosticSettings), location, database.id);
+                            } else {
+                                if (!diagnosticSettings.data.length) {
+                                    helpers.addResult(results, 2, 'diagnostic settings not configured for SQL database', location, database.id);
```
Comment (fatima99s) [inline plugins/azure/sqldatabases/dbDiagnosticLoggingEnabled.js:55]:
```suggestion
                                    helpers.addResult(results, 2, 'Diagnostic settings not configured for SQL database', location, database.id);
```

File: plugins/azure/sqldatabases/dbDiagnosticLoggingEnabled.js [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+                                } else { 
+                                    diagnosticSettings.data.forEach(settings=> { 
+                                        var enabledDiagnosticSettings = [...settings.metrics, ...settings.logs].filter((e => e.enabled)).map((e)=>e.category);
+                                        var skippedRecommendedSettings = recommendedDiagnosticSettings.filter((e) => !enabledDiagnosticSettings.includes(e));
+                                        if (skippedRecommendedSettings.length) {
+                                            helpers.addResult(results, 2, 'diagnostic settings are not configured with minimum requirements', location, settings.id);
```
Comment (fatima99s) [inline plugins/azure/sqldatabases/dbDiagnosticLoggingEnabled.js:61]:
```suggestion
                                            helpers.addResult(results, 2, 'Diagnostic settings are not configured with minimum requirements', location, settings.id);
```

File: plugins/azure/sqldatabases/dbDiagnosticLoggingEnabled.js [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+                                        } else {
+                                            helpers.addResult(results, 0,
+                                                'diagnostic settings are configured with minimum requirements', location, settings.id);
```
Comment (fatima99s) [inline plugins/azure/sqldatabases/dbDiagnosticLoggingEnabled.js:64]:
```suggestion
                                                'Diagnostic settings are configured with minimum requirements', location, settings.id);
```

**Review by mehakseedat63 — CHANGES_REQUESTED** [review]
We don't need to look for specific settings, should pass if any kind of diagnostic logs are enabled. Look at other diagnostic logging plugins for reference.

**Comment by muzzamilinovaqo** [comment]
> We don't need to look for specific settings, should pass if any kind of diagnostic logs are enabled. Look at other diagnostic logging plugins for reference.

The requirement is to have at least the mention settings enabled.

File: exports.js [hunk 5]
```diff
@@ -989,7 +990,7 @@ module.exports = {
         'eventHubMinimumTLSversion'     : require(__dirname + '/plugins/azure/eventhub/eventHubMinimumTLSversion.js'),
 
         'accessLogsEnabled'             : require(__dirname + '/plugins/azure/frontdoor/accessLogsEnabled.js'),
-        'frontDoorMinimumTlsVersion'    : require(__dirname + '/plugins/azure/frontdoor/frontDoorMinimumTlsVersion.js')
+        'frontDoorMinimumTlsVersion'    : require(__dirname + '/plugins/azure/frontdoor/frontDoorMinimumTlsVersion.js'),
```
Comment (alphadev4) [inline exports.js:993]:
```suggestion
        'frontDoorMinimumTlsVersion'    : require(__dirname + '/plugins/azure/frontdoor/frontDoorMinimumTlsVersion.js')
```

File: plugins/azure/sqldatabases/dbDiagnosticLoggingEnabled.js [hunk 6]
```diff
@@ -0,0 +1,80 @@
+var async = require('async');
+var helpers = require('../../../helpers/azure');
+
+module.exports = {
+    title: 'Database Diagnostic Logging Enabled',
```
Comment (alphadev4) [inline plugins/azure/sqldatabases/dbDiagnosticLoggingEnabled.js:5]:
```suggestion
    title: 'SQL Diagnostic Logging Enabled',
```

File: plugins/azure/sqldatabases/dbDiagnosticLoggingEnabled.js [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+    category: 'SQL Databases',
+    domain: 'Databases',
+    description: 'Ensure diagnostic logging is enabled for SQL databases.',
+    more_info: 'Enabling diagnostic logging provides valuable insights into SQL database performance and helps identify issues.',
+    recommended_action: 'Enable diagnostic logging for SQL databases with the minimum required data recording settings: SQLInsights, ErrorsTimeouts, BlocksDeadlocks, BasicInstanceAndApp, AdvancedWorkloadManagement.',
+    link: 'https://docs.microsoft.com/en-us/azure/sql-database/sql-database-monitoring-with-dmvs?tabs=sql-insights',
```
Comment (alphadev4) [inline plugins/azure/sqldatabases/dbDiagnosticLoggingEnabled.js:11]:
```suggestion
    link: 'https://learn.microsoft.com/en-us/azure/azure-sql/database/monitoring-sql-database-azure-monitor?view=azuresql',
```

File: plugins/azure/sqldatabases/dbDiagnosticLoggingEnabled.js [hunk 8]
```diff
@@ -0,0 +1,80 @@
+var async = require('async');
+var helpers = require('../../../helpers/azure');
+
+module.exports = {
+    title: 'Database Diagnostic Logging Enabled',
+    category: 'SQL Databases',
+    domain: 'Databases',
+    description: 'Ensure diagnostic logging is enabled for SQL databases.',
```
Comment (alphadev4) [inline plugins/azure/sqldatabases/dbDiagnosticLoggingEnabled.js:8]:
```suggestion
    description: 'Ensures diagnostic logging is enabled for SQL databases.',
```

File: plugins/azure/sqldatabases/dbDiagnosticLoggingEnabled.js [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+    more_info: 'Enabling diagnostic logging provides valuable insights into SQL database performance and helps identify issues.',
```
Comment (alphadev4) [inline plugins/azure/sqldatabases/dbDiagnosticLoggingEnabled.js:9]:
```suggestion
    more_info: 'Enabling diagnostic logging provides valuable insights into SQL database that helps to monitor resources for their availability, performance, and operation.',
```

File: plugins/azure/sqldatabases/dbDiagnosticLoggingEnabled.js [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+                    if (!databases.data.length) {
+                        helpers.addResult(results, 0,
+                            'No databases found for SQL server', location, server.id);
+                    } else {
+                        databases.data.forEach(database=> {
+                            
+                            var diagnosticSettings = helpers.addSource(cache, source, ['diagnosticSettings', 'listByDatabase', location, database.id]);
+                        
+                            if (!diagnosticSettings || diagnosticSettings.err || !diagnosticSettings.data) {
+                                helpers.addResult(results, 3, 'Unable to query diagnostic settings: ' + helpers.addError(diagnosticSettings), location, database.id);
```
Comment (alphadev4) [inline plugins/azure/sqldatabases/dbDiagnosticLoggingEnabled.js:52]:
```suggestion
                                helpers.addResult(results, 3, 'Unable to query SQL database diagnostic settings: ' + helpers.addError(diagnosticSettings), location, database.id);
```

**Comment by alphadev4** [comment]
> > We don't need to look for specific settings, should pass if any kind of diagnostic logs are enabled. Look at other diagnostic logging plugins for reference.
> 
> The requirement is to have at least the mention settings enabled.

We still cannot produce results based on settings.id ? We should filter out the missing settings from recommended settings and then produce the result accordingly

File: plugins/azure/sqldatabases/dbDiagnosticLoggingEnabled.js [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+                                    !logs_metrics.some(log => (log.category && log.category.toLowerCase() === requiredCategory && log.enabled) || log.categoryGroup === 'allLogs' && log.enabled)
+                                );
+                            });
+        
+                        }
+                        if (!missingConfig.length && found) {
+                            helpers.addResult(results, 0, 'SQL database has diagnostic logs enabled', location, database.id);
+        
+                        } else {
+                            helpers.addResult(results, 2, `SQL database does not have diagnostic logs enabled ${missingConfig.length ? `for following: ${missingConfig}` : ''}`, location, database.id);
```
Comment (mehakseedat63) [inline plugins/azure/sqldatabases/dbDiagnosticLoggingEnabled.js:88]:
```suggestion
                            helpers.addResult(results, 2, `SQL database does not have diagnostic logs enabled ${missingConfig.length ? `for following: ${missingConfig,join(','}` : ''}`, location, database.id);
```

**Review by mehakseedat63 — CHANGES_REQUESTED** [review]

File: plugins/azure/sqldatabases/dbDiagnosticLoggingEnabled.js [hunk 12]
```diff
@@ -0,0 +1,100 @@
+var async = require('async');
+var helpers = require('../../../helpers/azure');
+
+module.exports = {
+    title: 'Database Diagnostic Logging Enabled',
```
Comment (mehakseedat63) [inline plugins/azure/sqldatabases/dbDiagnosticLoggingEnabled.js:5]:
```suggestion
    title: 'SQL Database Diagnostic Logging Enabled',
```

File: plugins/azure/sqldatabases/dbDiagnosticLoggingEnabled.js [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+    category: 'SQL Databases',
+    domain: 'Databases',
+    description: 'Ensures diagnostic logging is enabled for SQL databases.',
+    more_info: 'Enabling diagnostic logging provides valuable insights into SQL database that helps to monitor resources for their availability, performance, and operation.',
+    recommended_action: 'Enable diagnostic logging for SQL databases with the minimum required data recording settings: SQLInsights, Errors, Timeouts, Blocks, Deadlocks, Basic, InstanceAndAppAdvanced, WorkloadManagement.',
```
Comment (mehakseedat63) [inline plugins/azure/sqldatabases/dbDiagnosticLoggingEnabled.js:10]:
```suggestion
    recommended_action: 'Enable diagnostic logging for all SQL databases.',
```

File: plugins/azure/sqldatabases/dbDiagnosticLoggingEnabled.js [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+    link: 'https://learn.microsoft.com/en-us/azure/azure-sql/database/monitoring-sql-database-azure-monitor?view=azuresql',
+    apis: ['servers:listSql', 'databases:listByServer', 'diagnosticSettings:listByDatabase'],
+    realtime_triggers: ['microsoftsql:servers:write', 'microsoftsql:servers:delete', 'microsoftsql:servers:databases:write', 'microsoftsql:servers:databases:delete','microsoftinsights:diagnosticsettings:write','microsoftinsights:diagnosticsettings:delete'],
+    settings: {
+        diagnostic_logs: {
```
Comment (mehakseedat63) [inline plugins/azure/sqldatabases/dbDiagnosticLoggingEnabled.js:15]:
```suggestion
        sqldb_diagnostic_logs: {
```
should make sure that this setting name is unique in all diagnostic log plugins

**Review by mehakseedat63 — CHANGES_REQUESTED** [review]

**Pull request merged**